### DISCUSSION TITLE
How to run two deployments on one computer

### DISCUSSION THREAD

**Discussion by Emil808** [nasa/fprime#1305] [Q&A] [answered]
I have two deployments `DepA` and `DepB`

DepA has one component that, when issued a command, will put a number to a port. 

DepB has one passive component that sends an event and updates telemetry when it receives a number from the port. 

The Deployments are connected with the generic hub with TCP. DepB uses the Drv/TcpServer, and DepA uses the Drv/TcpClient. The host address and port are hard-coded in the respective instances.fpp for now as `0.0.0.0` and `8080`

I did `fprime-gds -h` to see the options that I can change, but I am unsure of how to do it correctly or if its possible to two gds on one computer. 

- How do I correctly choose the address and ports for the GUI, TTS, IP Adapter Server, and hub connections? 
- How do I run the fprime-gds to have both running on my computer? 

Then when I try the same thing running on RPI, would all I need to do in main is to do the same as how the RPI connects to the gds, and add so that it takes in two more inputs for the address and port of the other Deployment?

**Comment by Emil808** [comment]
From investigating `fprime-gds -h` I figured out how to setup the GUI, TTS, and IP.

But now I'm having trouble with the TCP connections between the hubs. In the log files, 
`DepA` is reporting: `[WARNING] Failed to open port with status -4 and errno 111`
`DepA2` is reporting: `[WARNING] Failed to open port with status -11 and errno 9`

What am I missing to establish the TCP connection between the TCP server and TCP client that connects the generic hubs? 

In `DepA` (TCP Client) has: 
```
instance commP: Drv.ByteStreamDriverModel base id 0x5000 \
    at "../../Drv/TcpClient/TcpClient.hpp" \
    {

    phase Fpp.ToCpp.Phases.instances """
    Drv::TcpClient commP(FW_OPTIONAL_NAME("commP"));
    """

    phase Fpp.ToCpp.Phases.configConstants """
    enum {
        PRIORITY = 100,
        STACK_SIZE = Default::stackSize
    };
    """

    phase Fpp.ToCpp.Phases.startTasks """
    // Initialize socket server if and only if there is a valid specification
    const char * hostNameP = "127.0.5.0"; 
    U32 portNumberP = 50020; 
    if (hostNameP != nullptr && portNumberP != 0) {
        Os::TaskString namep("ReceiveTaskP");
        // Uplink is configured for receive so a socket task is started
        commP.configure(hostNameP, portNumberP);
        commP.startSocketTask(
            namep,
            true,
            ConfigConstants::commP::PRIORITY,
            ConfigConstants::commP::STACK_SIZE
        );
    }
    """

    phase Fpp.ToCpp.Phases.freeThreads """
    commP.stopSocketTask();
    (void) commP.joinSocketTask(nullptr);
    """

    }
```

`DepA2` (TCP Server) has: 
```
instance commP: Drv.ByteStreamDriverModel base id 0x5000 \
    at "../../Drv/TcpServer/TcpServer.hpp" \
    {

    phase Fpp.ToCpp.Phases.instances """
    Drv::TcpServer commP(FW_OPTIONAL_NAME("commP"));
    """

    phase Fpp.ToCpp.Phases.configConstants """
    enum {
        PRIORITY = 100,
        STACK_SIZE = Default::stackSize
    };
    """

    phase Fpp.ToCpp.Phases.startTasks """
    // Initialize socket server if and only if there is a valid specification
    const char * hostNameP = "127.0.5.0"; 
    U32 portNumberP = 50030; 
    if (hostNameP != nullptr && portNumberP != 0) {
        Os::TaskString namep("ReceiveTaskP");
        // Uplink is configured for receive so a socket task is started
        commP.configure(hostNameP, portNumberP);
        commP.startSocketTask(
            namep,
            true,
            ConfigConstants::commP::PRIORITY,
            ConfigConstants::commP::STACK_SIZE
        );
    }
    """

    phase Fpp.ToCpp.Phases.freeThreads """
    commP.stopSocketTask();
    (void) commP.joinSocketTask(nullptr);
    """

    }
```

**Reply by LeStarch** [reply]
First ensure that A2 is opening up a port properly.  Your IP for A2 seems wrong to me: `127.0.5.0` ends in a `0` which may be invalid.  Have you tried `127.0.0.1` which is the usual port for local host.  If this IP address is right, you need to figure out why you are getting an errno 9.

**Reply by Emil808** [reply]
In the startTask for the TCP Server, I forgot to add 
`commP.startup()`

**Reply by Emil808** [reply]
Now when I run A2 with `fprime-gds`, 
and run A with `fprime-gds --gui-port 5001 --tts-port 50051 --ip-port 50001 --ip-address 127.0.3.2 --tts-addr 127.0.3.2`

DepA with the Tcp Client connects to to address:port of the Tcp Server, but then shutsdown with: 
`[INFO] Shutting down F prime due to error. Failed to run DepA Application
Failed to open socket at 127.0.3.2:50001, retrying: OSError: [Errno 22] Invalid argument`

I found the same error being talked about with issue #1060. I don't know what to do from here.

**Reply by Emil808** [reply]
Here is the message after running the `fprime-gds <settings>` command
```
[INFO] Ensuring DepA Application is stable for at least 1 seconds
[INFO] Running Application: <>/fprime/DepA/build-artifacts/Linux/bin/DepA
[INFO] Log File: <>/fprime/DepA/logs/2022_03_07-15_53_17/DepA.log
Read failure attempting reconnection. ConnectionResetError: [Errno 104] Connection reset by peer
[ERROR] Failed to run application: <>/fprime/DepA/build-artifacts/Linux/bin/DepA -p 50001 -a 127.0.3.1. Error: <>/fprime/DepA/build-artifacts/Linux/bin/DepA stopped with code -11 sooner than 1 seconds.
```
then it spits out the .log file with the last line being: 
```
[INFO] Shutting down F prime due to error. Failed to run DepA Application Failed to open socket at 127.0.3.2:50001, retrying: OSError: [Errno 22] Invalid argument
```

**Reply by Emil808** [reply]
Changed the setup to not use "0.0.0.0" address, and the port numbers on both sides being the same resulted in the same error message.

**Reply by LeStarch** [reply]
There may be a bug in the GDS preventing custom ports. Let me check

**Reply by LeStarch** [reply]
Looking through this I don't see a specific error.  It seems the TTS port is failing to open claiming an invalid argument.  However, I ran your exact command line and it work correctly on my end.

The only thing I was able to see was this: https://stackoverflow.com/questions/30268008/udp-socket-sendto-functions/30268744#30268744.  This was linked as a possible source of errors with errno 22 when doing socket programing. In short, you should make sure to use the same network across the system.

On a single computer, you could start with a single ip address (e.g. the loopback address `127.0.0.1` for all your addresses).  Alternatively you can start with the "current network" address `0.0.0.0`.  Once you have one of those two working, try to switch to something like `127.1.2.3` assuming that is in your loop back range.

**Reply by Emil808** [reply]
Thanks for your help. I tried different addresses and port combinations. This is now becoming a Tcp socket programming problem than an Fprime problem. 

The Server deployment is set up with an address "127.0.5.1" and port 50020. The client dep is set up with "127.0.5.1" and port 50020. With this setup, the Server Dep runs and is listening for a single client at 127.0.5.1:50020. Then when the Client Dep runs and connects to the Server Dep, but the Client Dep reports a Socket_Failed_to_Accept error under the 1 second check and the Deployment attempts to restart, but the IP Adapter Server socket that was opened didn't close from the initial run and after the restart, the application attempted to open it again but fails.

When I change the port on the Client Dep from 50020 to 50021 to be different to what was chosen on the Server Dep, both applications end up running to when the gds-gui opens and I can see the health telemetry from both. The log file for Server Dep reports it is listening for a single client, but the Client Dep is reporting Sock_Failed_to_Connect.

**Reply by Emil808** [reply]
I did a print out to see which address:port was the one failing, and it is the ip-address and ip-port that is failing. Is there some conflict with opening running two fprime-gds commands on one machine at the same time?

**Reply by LeStarch** [reply]
If you use the same --ip-address and --ip-port arguments, yes.  Each invocation of the comm layer (talks to FSW) needs a dedicated port to run on. 

However, I am confused, aren't you using one deployment of FSW as a hub to the other? If so, you'd just connect to the one deployment?

**Reply by LeStarch** [reply]
How is this system intended to be setup?

Client Dep -> Server Dep -> GDS -> Browser

Or

Client Dep -> GDS 1 -> Browser 1
Server Dep -> GDS 2 -> Browser 2

Something else?

When connection to a deployment the GDS launches a TCP server @ (ip-address, ip-port) and the FSW needs a  TCP client to connect to it.  If you are using the TcpServer component, it needs to operate on a different port from the (ip-address, ip-port) specified to the GDS or the two will collide (both opening a servier on that addr/port).

Typically when using the GenericHub we do the following:

Client Dep -> (Sever Dep's server) -> GenericHub -> Server Dep's client -> GDS server -> Browser.

This means the server dep needs a server component facing the client dep, and a client facing the GDS.  It acts as a router for client deps' data but still needs a connection to the GDS.

**Reply by Emil808** [reply]
What I'm trying to implement the hub pattern shown [here](https://nasa.github.io/fprime/UsersGuide/best/hub-pattern.html)

The current goal is to have a component on Client Dep that will produce a number on command. That number will be transported to Server Dep, which has a component that will give out an event that says it has the number. Using browser 1 to issue a command should result in Browser 2 displaying an event. 

I have this to see each deployment's activity. 
Client Dep (client to gds)   -> GDS1 -> Browser 1
Server Dep (client to gds) -> GDS2  -> Browser 2

Then I have:
Client Dep (client to other)  -> Server Dep (server to other) 

I run each fprime-gds command with a different set of gui, tts, and ip addresses and ports.

**Reply by LeStarch** [reply]
If it is all running one one machine, for testing purposes, I'd use the same address `127.0.0.1` for all addresses and then use a separate port for each:  (ip port 1, ip port 2, tts-port 1, tts-port2, gui port 1, gui port 2).

In addition confirm that you have 4 Tcp* components, 2 in each deployment:

1. TcpClient (dep 1) with same address and port at ip port 1 in GDS 1 above
2. TcpClient (dep 2) with same address and port as ip port 2 in GDS 2 above
3. TcpClient (dep 1) with same address and port at TcpServer in dep2
4. TcpServer (dep 2) with same address and port ad TcpClient in dep1

Make sure framer/deframer are wired to item 1 or 2 above.  Cross wire the hub  on items 3/4.

I might also comment out the hub for now and verify that you can get 2 deployments producing separate telemetry through two GDSs (item 1, 2).  Once that works reliably, add in items 3, 4.

I hope this is helping debug.

**Reply by Emil808** [reply]
With your suggestions, having item 1, 2 run without the hubs works. Both browsers open and I can get changing telemetry from both. 

When adding the hubs, I first issue fprime-gds for TcpServer, followed by TcpClient. TcpClient gives this error: 
`Read failure attempting reconnection. ConnectionResetError: [Errno 104] Connection reset by peer
[ERROR] Failed to run application: <>prime/DepA/build-artifacts/Linux/bin/DepA -p 60001 -a 0.0.0.0. Error:<>/fprime/DepA/build-artifacts/Linux/bin/DepA stopped with code -11 sooner than 1 seconds.`
...
`[INFO] Shutting down F prime due to error. Failed to run DepA Application
Failed to open socket at 0.0.0.0:60001, retrying: OSError: [Errno 22] Invalid argument
`

If I run TcpClient first, then TcpServer, TcpClient Console output gives when the hub client connects: 
```
Read failure attempting reconnection. ConnectionResetError: [Errno 104] Connection reset by peer
```
The browser for TcpClient comes up, and attempting to issuing a command give this console output: 
`Uplink failed to send 24 bytes of data after 3 retries`

So from this, when TcpClient hub-client connects to TcpServer hub-server, this causes a read failure and connection reset on the comm side within the fprime-gds portion since it is coming out of the console. No error or info on the log file, but the hub-client might be sending something to the fprime-gds portion along side sending to the hub-server.

**Reply by Emil808** [reply]
I've attempted this setup by running each deployment on a different computer and still getting the same problem. 

I added a connection to the client Comm component to allocate with static memory and that avoided the client from closing instantly. So is it how I connected the hubs to other components?

Here is DepA with the producer Component -> hub -> framer (downlinkP -> commP (client)
```
connections hub{
            SimpleProducer.valOut -> hub.portIn[0]
            
            hub.dataOutAllocate -> staticMemory.bufferAllocate[Ports_StaticMemory.hub]
            hub.dataOut -> downlinkP.bufferIn
            downlinkP.bufferDeallocate -> staticMemory.bufferDeallocate[Ports_StaticMemory.hub]

            downlinkP.framedAllocate -> staticMemory.bufferAllocate[Ports_StaticMemory.downlinkP]
            downlinkP.framedOut -> commP.send
            commP.deallocate -> staticMemory.bufferDeallocate[Ports_StaticMemory.downlinkP]
            
            commP.allocate ->staticMemory.bufferAllocate[Ports_StaticMemory.hub]
        }
```

DepA2 with CommP (server) -> deframer (uplinkP) -> hub -> receiver component:
```
connections HubUplink{
            commP.allocate -> staticMemory.bufferAllocate[Ports_StaticMemory.commP]
            commP.$recv -> uplinkP.framedIn
            uplinkP.framedDeallocate -> staticMemory.bufferDeallocate[Ports_StaticMemory.commP]

            uplinkP.bufferAllocate -> staticMemory.bufferAllocate[Ports_StaticMemory.hub]
            uplinkP.bufferOut -> hub.dataIn
            uplinkP.bufferDeallocate -> staticMemory.bufferDeallocate[Ports_StaticMemory.hub]
            
            hub.portOut[0] -> SimpleReceiver.valIn
        }
```

**Comment by Emil808** [comment, marked as answer]
The Deployment with the TcpClient, is being used to _only_ send data down, never receive. So it did not need to run the StartSocketTask that runs the SocketReadTask. This read task is useless for this application. It was crashing at a getBuffer invocation to an allocate port that wasn't connected, which caused a FW_ASSERT which caused a segmentation fault because `vfprintf_internal.c does not exist` (Another problem that to be investigated later). 

Instead of doing the suggested in instances.fpp: 
```c++
const char * hostNameP = "0.0.0.0"; 
    U32 portNumberP = 50020; 
    if (hostNameP != nullptr && portNumberP != 0) {
        Os::TaskString name("ReceiveTaskP");
        // Uplink is configured for receive so a socket task is started
        commP.configure(hostNameP, portNumberP); 
         commP.startSocketTask(name);
         commP.stopSocketTask();
         (void) commP.joinSocketTask(nullptr);
    }
    """

    phase Fpp.ToCpp.Phases.freeThreads """
    commP.close(); 
    """
```
I just open and close (this requires the TcpServer to be up and running first): 
```c++
const char * hostNameP = "0.0.0.0"; 
    U32 portNumberP = 50020; 
    if (hostNameP != nullptr && portNumberP != 0) {
        Os::TaskString name("ReceiveTaskP");
        // Uplink is configured for receive so a socket task is started
        commP.configure(hostNameP, portNumberP); 
        commP.open(); 
    }
    """

    phase Fpp.ToCpp.Phases.freeThreads """
    commP.close(); 
    """
```
I have two Deployments, DepA (with a TcpClientComponent), DepA2 (with a TcpServer Component), and generic hubs to bridge a producer and receiver component.
